apps/language_server/lib/language_server/providers/code_mod/ast.ex

```diff
@@ -0,0 +1,29 @@
+defmodule ElixirLS.LanguageServer.Providers.CodeMod.Ast do
+  alias ElixirLS.LanguageServer.SourceFile
+
+  @type source :: SourceFile.t() | String.t()
+  @type t ::
+          atom()
+          | binary()
+          | [any()]
+          | number()
+          | {any(), any()}
+          | {atom() | {any(), [any()], atom() | [any()]}, Keyword.t(), atom() | [any()]}
+
+  @spec from(source() | String.t()) :: {:ok, t()} | :error
+  def from(%SourceFile{text: text}) do
+    from(text)
+  end
+
+  def from(text) when is_binary(text) do
+    case ElixirSense.string_to_quoted(text, {1, 1}) do
+      {:ok, ast} -> {:ok, ast}
+      _ -> :error
+    end
+  end
```

Comment:
The following test of unused variables would not pass:
```elixir
test "applied to a branch in a case" do
  {:ok, result} =
	~q[
	  case my_thing do
		{:ok, unused} -> :ok
		_ -> :error
	  end
	]
	|> modify(line: 1)

  expected = ~q[
	case my_thing do
	  {:ok, _unused} -> :ok
	  _ -> :error
	end
  ]

  assert result == expected
end
```
because `ElixirSense.string_to_quoted("  {:ok, unused} -> :ok", {1, 1})` returns 
```elixir
{:ok, {:__atom_elixir_marker_1__, [closing: [line: 1, column: 27], line: 1, column: 2], []}}
```
that is not an AST of that line we can use to add underscore. It seems like ElixirSense issue. Now, in that case, we are just not producing any code action.

Reply:
You cannot use `ElixirSense` here. Better stick with `Code.string_to_quoted_with_comments` as it preserves more info and do not attempt to fix errors

Reply:
I see. But it seems that `Code.string_to_quoted_with_comments` was added in Elixir 1.13 and ElixiLS supports 1.12. Will that be a problem?

Reply:
Not really. For now we can limit this feature to 1.13+ and 1.12 goes out of support in few months anyway